Refactoring serving model and eval model export directory structure

18jeffreyma opened the pull request:
For Estimator-based models, models are currently not exported to the root directory, but to a deeper timestamped folder structure. Since the ModelRun artifact now contains all non model export related files, the Model artifact directory now is automatically cleaned to only contain the exported model in the base directory. 

Previous Structure for Estimator-based model exports:

   |-- ModelExportPath
       |---- EVAL_MODEL_DIR  <- eval_model_dir
           |------ timestamped model  <- eval_model_path
               |-------- saved_model.pb
               |-------- ...
       |---- SERVING_MODEL_DIR  <- serving_model_dir
           |------ export
               |-------- exporter name
                   |---------- timestamped model  <- serving_model_path
                       |------------ saved_model.pb
                       |------------ ...

Refactored structure:

   |-- ModelExportPath
       |---- EVAL_MODEL_DIR  <- eval_model_dir
           |------ saved_model.pb
           |------ ...
       |---- SERVING_MODEL_DIR  <- serving_model_dir
           |------ saved_model.pb
           |------ ...

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -23,47 +23,39 @@
+    return serving_model_path(output_uri)
 
 
 def serving_model_dir(output_uri: Text) -> Text:
```
let's just deprecate this serving_model_dir & eval_model_dir function

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -58,6 +59,31 @@ def _is_chief():
 
+def _serving_model_path(working_dir: Text) -> Text:
+  """Given model artifact working directory, returns original path
+  for timestamped and named serving model exported."""
```
less than 80 chars, otherwise you need a full doc format like  [1], same else where

[1]https://github.com/tensorflow/tfx/blob/master/tfx/components/example_gen/base_example_gen_executor.py#L175

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -166,6 +192,46 @@ def _GetFnArgs(self, input_dict: Dict[Text, List[types.Artifact]],
         **custom_config)
 
+
+  def _CleanServingDir(self, working_dir: Text) -> None:
```
let's just pass modelrun path to fn.serving_model_dir (add some comment), and copy the model out of modelrun to model artifact

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -166,6 +192,46 @@ def _GetFnArgs(self, input_dict: Dict[Text, List[types.Artifact]],
+
+    if model_export_dir != serving_model_dir:
+      # Copy serving model export to base serving model directory
+      tempdir = tempfile.mkdtemp()
```
use self.working_dir(), it will create a temp path under pipeline_root.

I'm not sure if temfile works on Cloud or other file systems

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -219,6 +285,12 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
 
+    # Clean model export directories.
+    working_dir = artifact_utils.get_single_uri(
+        output_dict[constants.MODEL_KEY])
```
ModelRun artifact should be the working dir

18jeffreyma commented:
Implemented approach 2 (as discussed with Goutham). For Generic Trainer, there is no change in behavior. For Estimator based trainer, the Executor gives the user a modified set of `fn_args` with the `serving_model_dir` and `eval_model_dir` inside the `model_run` directory. Also updated examples/test cases which used `keras_to_estimator`.

1025KB commented on tfx/examples/iris/iris_utils.py:
```diff
@@ -206,23 +208,46 @@ def run_fn(fn_args: executor.TrainerFnArgs):
+  fn_args_dict['eval_model_dir'] = os.path.join(fn_args.model_run_dir,
+                                                path_utils.EVAL_MODEL_DIR)
+  user_fn_args = executor.TrainerFnArgs(arg_dict=fn_args_dict)
+  training_spec = trainer_fn(user_fn_args, schema)
```
trainer_fn is not used by tfx (it should be _trainer_fn) in this example, you can direct update trainer_fn

1025KB commented on tfx/examples/iris/iris_utils.py:
```diff
@@ -206,23 +208,46 @@ def run_fn(fn_args: executor.TrainerFnArgs):
+  eval_source = executor.eval_model_path(fn_args.model_run_dir)
+  eval_dest = fn_args.eval_model_dir
+  io_utils.copy_dir(eval_source, eval_dest)
+  absl.logging.info('Eval model copied to: %s.', eval_dest)
```
duplicated code, considering create a util function

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -23,60 +23,36 @@
-
-def eval_model_dir(output_uri: Text) -> Text:
-  """Returns directory for exported model for evaluation purpose."""
-  return os.path.join(output_uri, EVAL_MODEL_DIR)
```
my bad, I think we still need this eval_model_dir and serving_model_dir 

so everything related to path is controlled in this file

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -58,18 +58,46 @@ def _is_chief():
   return task_type == 'chief' or (task_type == 'master' and task_index == 0)
 
 
+def serving_model_path(working_dir: Text) -> Text:
```
we need this in the path_utils for back compat, check my comments below

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -23,60 +23,36 @@
+    return serving_model_path(output_uri)
 
 
 def serving_model_path(output_uri: Text) -> Text:
```
ideally we want the new version but we still need this for back compat I think

we want sth like
```
serving_model_path():
if old structure:
  log warning, mention we are going to deprecate the support of this structure, please have xxx structure
  return model path
if new structure:
  return model path
```

18jeffreyma commented:
Addressed above comments. Some notable changes: 
* TrainerFnArgs has same interface, but had to have it inherit from `dict` to achieve the attribute/key assignment
* Undeprecated serving dir and eval dir methods, and moved all path utilities to path_utils
* Added a dir_utils file containing a method for copying models to model artifact

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -302,12 +313,14 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
+
     else:
       absl.logging.info(
           'eval_savedmodel export for TFMA is skipped because '
```
update this 

change to model export is skipped ....

1025KB replied:
I mean just leave one log info

"model export is skipped because this is not the chief worker"

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -115,7 +115,7 @@ def _GetFnArgs(self, input_dict: Dict[Text, List[types.Artifact]],
     # TODO(ruoyu): Make this a dict of tag -> uri instead of list.
     if input_dict.get(constants.BASE_MODEL_KEY):
-      base_model = path_utils.serving_model_path(
+      base_model = path_utils.serving_model_working_path(
```
let's keep the old name otherwise there is a lot place need to be changed

1025KB commented on tfx/examples/iris/iris_utils.py:
```diff
@@ -206,6 +208,16 @@ def run_fn(fn_args: executor.TrainerFnArgs):
+  # Modify fn_args with model_run_dir given as the working directory.
+  # Executor will then copy user models to model artifact directory.
+  serving_dest = fn_args.serving_model_dir
+  eval_dest = fn_args.eval_model_dir
```
change to _trainer_fn, and do it like testdata/module file (avoid updating fn_args in module file)

1025KB replied:
unsolved?

1025KB commented on tfx/examples/iris/iris_utils_native_keras.py:
```diff
@@ -263,4 +263,5 @@ def run_fn(fn_args: TrainerFnArgs):
                                             dtype=tf.string,
                                             name='examples')),
   }
+  absl.logging.warning(fn_args.serving_model_dir)
```
forget to remove?

1025KB replied:
let's mark it resolve after fix (same elsewhere)

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -28,42 +29,36 @@
-# TODO(b/127149760): simplify this PPP-esque structure.
-#
-# Directory structure of exported model for estimator based trainer:
-#   |-- <ModelExportPath>
```
mention the back compat for this structure and add a todo for deprecation

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -28,42 +29,36 @@
 
 def eval_model_path(output_uri: Text) -> Text:
-  """Returns path to timestamped exported model for evaluation purpose."""
+  """Returns final path to exported model for evaluation purpose."""
```
merge eval_model_working_path into this eval_model_path (will be `if` `elif` + `else`)

as downstream still call eval_model_path, we want downstream to work with old version (e.g., a model artifact created by previous tfx version is resolved by Resolver to use in current pipeline)

18jeffreyma replied:
since eval model is timestamped, i've just replaced it with a try catch (i.e. try to get path and except if it doesnt exist).

1025KB replied:
you can check if saved_model.pb exists in model_dir, if so return model_dir else return io_utils.get_only_uri_in_dir(model_dir)

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -72,11 +67,32 @@ def serving_model_dir(output_uri: Text) -> Text:
 
 
 def serving_model_path(output_uri: Text) -> Text:
```
merger serving_model_working_path in to serving_model_path

the functionality for serving_model_path should be `get the pure model path no matter the structure`

18jeffreyma commented on tfx/components/trainer/executor.py:
```diff
@@ -58,18 +59,17 @@ def _is_chief():
+    if key in self:
+      return self[key]
+    else:
+      raise AttributeError('No such attribute: ' + key)
```
yep, also had some weird recursion depth issues with having both getitem and getattr, so I just changed it to a dict

18jeffreyma commented:
Addressed comments above, thanks again!

1025KB commented on tfx/components/testdata/module_file/trainer_module.py:
```diff
@@ -338,11 +339,17 @@ def run_fn(fn_args: executor.TrainerFnArgs):
 
-  absl.logging.info('Exported eval_savedmodel to %s.', fn_args.eval_model_dir)
+  absl.logging.info('Training complete. Model written to %s',
+                    fn_args.serving_model_dir)
```
duplicated, remove 333-334

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -302,14 +312,18 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
       absl.logging.info('Exported eval_savedmodel to %s.',
                         fn_args.eval_model_dir)
+
+      # TODO(b/160795287): Deprecate estimator based executor.
```
just keep line 283, remove this duplicated todo

18jeffreyma commented:
Addressed above issues. Had to revert some files back to original commits, which led to some previous issues becoming unresolved again, but should be resolved now.

1025KB commented on tfx/components/trainer/executor.py:
```diff
@@ -302,14 +312,13 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
-          'eval_savedmodel export for TFMA is skipped because '
-          'this is not the chief worker.'
-      )
+          'model export for TFMA is skipped because this is '
```
not just for TFMA, the eval is for TFMA

serving model export is also disabled for non chief

1025KB commented on tfx/utils/path_utils.py:
```diff
@@ -72,11 +95,32 @@ def serving_model_dir(output_uri: Text) -> Text:
+
+
+# TODO(b/160795287): Deprecate estimator based executor.
+def copy_model(working_dir: Text, dest: Text, tag: Text) -> None:
```
I feel this function only change 2 lines[1] of code to 1 lines[2], but the tag is a risky (you can pass the function in this function directly) and the readability is not as good as the two line solution (drop function and inline the code), wdyt?

[1]
source = path_fn(working_dir)
  io_utils.copy_dir(source, dest)
[2]
copy_model

18jeffreyma replied:
agreed, would make readability much easier and will remove need for extra if statement. will fix

18jeffreyma replied:
I'm leaning towards keeping the copy_model fn (since it does somewhat improve readability). Additionally, it also links the executor to the path_utils file (which describes the folder structure deprecation in more detail), in case users have difficulty understanding what's needed for the deprecation

1025KB replied:
I mean we can just do [1] in place instead of call this function, it also includes path_utils

another option is [2]

[1]
source = path_utils.serving_model_path(xxx)
io_utils.copy_dir(source, dest)

[2] 
def copy_model(working_dir, dest, model_extract_function):
   ...

then 
copy_model(working_dir, dest, path_utils.serving_model_path) thus you don't need to manually write the tag.

18jeffreyma replied:
gotcha, the only two reasons for the tag was (1) choosing path_fn (which is now just passed in) and (2) for the log message to be accurate. If we just do it in-line both of these issues can be resolved

18jeffreyma replied:
i think in-place might be preferable, will switch it to that

1025KB commented on tfx/components/testdata/module_file/trainer_module.py:
```diff
@@ -329,20 +330,27 @@ def run_fn(fn_args: executor.TrainerFnArgs):
+  absl.logging.info('Training complete. Model written to %s',
+                    fn_args.serving_model_dir)
+  absl.logging.info('Exported eval_savedmodel to %s.',
+                    fn_args.eval_model_dir)
```
this log is kind of duplicated (line 351 347, just keep one of them)

18jeffreyma commented:
Path util tests haven't been updated yet, will add them as soon as possible.
